RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc

```diff
@@ -564,23 +597,38 @@ void MuonIdProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
              newMuon = false;
              muon.setMatches( trackerMuon.matches() );
              if (trackerMuon.isTimeValid()) muon.setTime( trackerMuon.time() );
-             if (trackerMuon.isEnergyValid()) muon.setCalEnergy( trackerMuon.calEnergy() );
+             if (fillEnergy_ && trackerMuon.isEnergyValid()) muon.setCalEnergy( trackerMuon.calEnergy() );
              if (goodTrackerMuon) muon.setType( muon.type() | reco::Muon::TrackerMuon );
              if (goodRPCMuon) muon.setType( muon.type() | reco::Muon::RPCMuon );
+	     if (doME0_){
+	       if (goodME0Muon) muon.setType( muon.type() | reco::Muon::ME0Muon );
+	     }
              LogTrace("MuonIdentification") << "Found a corresponding global muon. Set energy, matches and move on";
              break;
            }
          }
          if ( newMuon ) {
-           if ( goodTrackerMuon || goodRPCMuon ){
-             outputMuons->push_back( trackerMuon );
-           } else {
-             LogTrace("MuonIdentification") << "track failed minimal number of muon matches requirement";
-             const reco::CaloMuon& caloMuon = makeCaloMuon(trackerMuon);
-             if ( ! caloMuon.isCaloCompatibilityValid() || caloMuon.caloCompatibility() < caloCut_ || caloMuon.p() < minPCaloMuon_) continue;
-             caloMuons->push_back( caloMuon );
-           }
-         }
+	   if (doME0_){
+	     if ( goodTrackerMuon || goodRPCMuon || goodME0Muon){
```

Comment:
Unless I'm missing something, I think the duplicate code lines 621-628 can be eliminated by condensing lines 611 & 612:
`if (goodTrackerMuon || goodRPCMuon || (doMeO_ && goodME0Muon)) {`